Dummy Data
```{r}
dummy_data_cosia <- readr::read_csv("~/Downloads/dummy_data_cosia.csv")
library(tidyverse)
dummy_data_cosia<-dummy_data_cosia %>% remove_rownames %>% column_to_rownames(var="...1")
dummy_data_cosia<-data.frame(t(dummy_data_cosia))
dummy_data_cosia<- dummy_data_cosia%>% mutate(gene1 = (gene1)^2 ,
                                              gene2 = (gene2)^2,
                                              gene3 = (gene3)^2,
                                              gene4 = (gene4)^2,
                                              gene5 = (gene5)^2
                                              )
library(BioQC)
DENTROPY_DIVERSITY_G<-data.frame(entropyDiversity(dummy_data_cosia,norm = TRUE)) # across genes
DENTROPY_DIVERSITY_T<-data.frame(entropyDiversity(t(dummy_data_cosia),norm = TRUE)) # across tissue
DENTROPY_SPECIFICITY_G<-data.frame(entropySpecificity(t(dummy_data_cosia),norm = TRUE)) # across genes
DENTROPY_SPECIFICITY_T<-data.frame(entropySpecificity(dummy_data_cosia,norm = TRUE)) # across tissue



```

```{r}
filter_mouse<- dplyr::select(mouse_specific, Anatomical.entity.name, VST, Gene.ID)
Mouse<-filter_mouse %>% group_by(Gene.ID,Anatomical.entity.name) %>% summarise(Sample_size = n(), Median_VST = median(VST), VST = paste(unique(VST), collapse = ', '))
Mouse<- dplyr::select(Mouse, Anatomical.entity.name, Median_VST, Gene.ID)
filter<- Mouse%>% tidyr::pivot_wider(names_from = Gene.ID, values_from = Median_VST )
filter <- filter %>% remove_rownames %>% tibble::column_to_rownames(var="Anatomical.entity.name")

```



```{r}
kidney <- CoSIAn(
gene_set = c("PKD1","PKD2","SETBP1"),
i_species="h_sapiens",
o_species=c("m_musculus", "h_sapiens"),
input_id="Symbol",
output_ids= "Ensembl_id",
mapping_tool="annotationDBI",
ortholog_database = "HomoloGene",
map_tissues = "",
map_species = "",
metric_type=""
)
kidney_ensembl<-getConversions(kidney)
k<-kidney_ensembl@converted_id

mouse_ensembl<-k$m_musculus_ensembl_id
```


Entropy and Specificity READ Counts
```{r}
filter_species <- dplyr::filter(mouse_specific,Gene.ID %in% mouse_ensembl)
filter_gex<- dplyr::select(filter_species, Anatomical.entity.name, VST, Gene.ID)
filter_gex$VST_Scaled_Exp<- 1 / (1 + exp(-filter_gex$VST))

library(dplyr)
library(magrittr)
library(tibble)


Species<-filter_gex %>% group_by(Gene.ID,Anatomical.entity.name) %>% summarise(Sample_size = n(), Median_VST = median(VST_Scaled_Exp), VST = paste(unique(VST_Scaled_Exp), collapse = ', '))
Species<- dplyr::select(Species, Anatomical.entity.name, Median_VST, Gene.ID)

filter_gene<- Species%>% tidyr::pivot_wider(names_from = Gene.ID, values_from = Median_VST )
filter_gene <- filter_gene %>% remove_rownames %>% tibble::column_to_rownames(var="Anatomical.entity.name")


filter_gex$ENSMUSG00000034462_L10<-log10(filter_gex$ENSMUSG00000034462)
filter_gex$ENSMUSG00000034462_L10L2<-log2(filter_gex$ENSMUSG00000034462)


```

```{r}
atomEntropy <- function(x) ifelse(x==0, 0, x*log2(x))
atomDiversity <- function(x) {
  rel <- x/sum(x, na.rm=TRUE)
  -sum(atomEntropy(rel))
}

#' Shannon entropy
#' 
#' @param vector A vector of numbers, or characters. Discrete probability of each item is calculated and the Shannon entropy is returned.
#' @return Shannon entropy
#' 
#' Shannon entropy can be used as measures of gene expression
#' specificity, as well as measures of tissue diversity and
#' specialization. See references below.
#' 
#' We use \code{2} as base for the entropy calculation, because in this
#' base the unit of entropy is \emph{bit}.
#'
#' @references   Martinez and Reyes-Valdes (2008) Defining diversity, 
#' specialization, and gene specificity in transcriptomes through information 
#' theory. PNAS 105(28):9709--9714
#'  
#' @author Jitao David Zhang <jitao_david.zhang@roche.com>
#' 
#' @examples 
#' myVec0 <- 1:9
#' entropy(myVec0) ## log2(9)
#' myVec1 <- rep(1, 9)
#' entropy(myVec1)
#' 
#' entropy(LETTERS)
#' entropy(rep(LETTERS, 5))
#' @export
entropy <- function(vector) {
  -sum(atomEntropy(table(vector)/length(vector)))
}

#' Entropy-based gene-expression specificity
#' @param mat A matrix (usually an expression matrix), with genes (features) in rows and samples in columns.
#' @param norm Logical, whether the specificity should be normalized by \code{log2(ncol(mat))}.
#' @return A vector of the length of the row number of the input matrix, namely the specificity score of genes.
#' 
#' @references   Martinez and Reyes-Valdes (2008) Defining diversity, 
#' specialization, and gene specificity in transcriptomes through information 
#' theory. PNAS 105(28):9709--9714
#' 
#' @seealso \code{\link{entropy}}
#' @examples
#' myMat <- rbind(c(3,4,5),c(6,6,6), c(0,2,4))
#' entropySpecificity(myMat)
#' entropySpecificity(myMat, norm=TRUE)
#'
#' myRandomMat <- matrix(runif(1000), ncol=20)
#' entropySpecificity(myRandomMat)
#' entropySpecificity(myRandomMat, norm=TRUE)
#' @export
EntropySpecificity <- function(mat, norm=FALSE) {
  mat.rel <- apply(mat, 2L, function(x) x/sum(x, na.rm=TRUE))
  speci <- apply(mat.rel, 1L, function(x) {
    gm <- mean(x, na.rm=TRUE)
    rel <- x/gm
    mean(atomEntropy(rel), na.rm=TRUE)
  })
  if(norm)
    speci <- speci/log2(ncol(mat))
  return(speci)
}

#' Entropy-based sample diversity
#' @param mat A matrix (usually an expression matrix), with genes (features) in rows and samples in columns.
#' @param norm Logical, whether the diversity should be normalized by \code{log2(nrow(mat))}.
#' @return A vector as long as the column number of the input matrix
#' 
#' @references   Martinez and Reyes-Valdes (2008) Defining diversity, 
#' specialization, and gene specificity in transcriptomes through information 
#' theory. PNAS 105(28):9709--9714
#' 
#' @seealso \code{\link{entropy}} and \code{\link{sampleSpecialization}}
#' @examples
#' myMat <- rbind(c(3,4,5),c(6,6,6), c(0,2,4))
#' entropyDiversity(myMat)
#' entropyDiversity(myMat, norm=TRUE)
#'
#' myRandomMat <- matrix(runif(1000), ncol=20)
#' entropyDiversity(myRandomMat)
#' entropyDiversity(myRandomMat, norm=TRUE)
#' @export
EntropyDiversity <- function(mat, norm=FALSE) {
  res <- apply(mat, 2L, atomDiversity)
  if(norm)
    res <- res/log2(nrow(mat))
  res
}


#' Entropy-based sample specialization
#' @param mat A matrix (usually an expression matrix), with genes (features) in rows and samples in columns.
#' @param norm Logical, whether the specialization should be normalized by \code{log2(ncol(mat))}.
#' @return A vector as long as the column number of the input matrix
#' 
#' @references   Martinez and Reyes-Valdes (2008) Defining diversity, 
#' specialization, and gene specificity in transcriptomes through information 
#' theory. PNAS 105(28):9709--9714
#' 
#' @seealso \code{\link{entropy}} and \code{\link{entropyDiversity}}
#' @examples
#' myMat <- rbind(c(3,4,5),c(6,6,6), c(0,2,4))
#' sampleSpecialization(myMat)
#' sampleSpecialization(myMat, norm=TRUE)
#'
#' myRandomMat <- matrix(runif(1000), ncol=20)
#' sampleSpecialization(myRandomMat)
#' sampleSpecialization(myRandomMat, norm=TRUE)
#' @export
sampleSpecialization <- function(mat, norm=TRUE) {
  speci <- entropySpecificity(mat, norm=FALSE)
  rel <- apply(mat, 2L, function(x) x/sum(x, na.rm=TRUE))
  res <- apply(rel, 2L, function(x) sum(x*speci, na.rm=TRUE))
  if(norm)
    res <- res/log2(ncol(mat))  
  res
}
```


```{r}
library(BioQC)
ENTROPY_DIVERSITY_G<-data.frame(entropyDiversity(filter_gene,norm = TRUE)) # across genes
ENTROPY_DIVERSITY_T<-data.frame(entropyDiversity(t(filter_gene),norm = TRUE)) # across tissue
ENTROPY_SPECIFICITY_G<-data.frame(entropySpecificity(t(filter_gene),norm = TRUE)) # across genes
ENTROPY_SPECIFICITY_T<-data.frame(entropySpecificity(filter_gene,norm = TRUE)) # across tissue
```

Entropy and Specificity VST
```{r}
t<-CoSIA::getTissues("m_musculus")
kidney <- CoSIA::CoSIAn(
gene_set = c("PKD1","PKD2","SETBP1"),
i_species="h_sapiens",
o_species=c("m_musculus", "h_sapiens"),
input_id="Symbol",
output_ids= "Ensembl_id",
mapping_tool="annotationDBI",
ortholog_database = "HomoloGene",
map_tissues = "kidney",
map_species = "m_musculus",
metric_type="DS_Gene"
)
kidney_ensembl<-CoSIA::getConversions(kidney)
kidney_gex<-CoSIA::getGEx(kidney_ensembl)
kidney_gex_metric<-CoSIA::getGExMetrics(kidney_gex)

mouse_ensembl<-kidney_ensembl$m_musculus_ensembl_id
```

```{r}
cosia_object <- CoSIAn(
    gene_set = "8452",
    i_species="h_sapiens",
    o_species="h_sapiens",
    input_id="Entrez_id",
    o_ids=c("Ensembl_id","Gene_name","Entrez_id"),
    mapping_tool="biomaRt",
    ortholog_database = "HomoloGene",
    map_tissues = "",
    map_species = "",
    metric_type=""
  )
getConversions(cosia_object)@converted_id
```

```{r}
CoSIAn_Obj <- CoSIA::CoSIAn(gene_set = FullHumanMDLKD$Human_Symbol,
                            i_species = "h_sapiens",
                            o_species = c("h_sapiens", "m_musculus", "r_norvegicus"),
                            input_id = "Symbol",
                            output_ids = "Ensembl_id",
                            map_species = c("h_sapiens", "m_musculus", "r_norvegicus"),
                            map_tissues = c("adult mammalian kidney", "heart"),
                            mapping_tool = "annotationDBI",
                            ortholog_database = "HomoloGene",
                            metric_type = "CV_Species"
                            )

CoSIAn_Obj_convert <- CoSIA::getConversions(CoSIAn_Obj)
```

